crates/ide_db/src/search.rs

```diff
@@ -352,14 +361,22 @@ impl<'a> FindUsages<'a> {
         };
 
         let pat = name.as_str();
+        let search_for_self = self.include_self_kw_refs;
+
         for (file_id, search_range) in search_scope {
             let text = sema.db.file_text(file_id);
             let search_range =
                 search_range.unwrap_or_else(|| TextRange::up_to(TextSize::of(text.as_str())));
 
             let tree = Lazy::new(|| sema.parse(file_id).syntax().clone());
 
-            for (idx, _) in text.match_indices(pat) {
+            let matches = text.match_indices(pat).chain(if search_for_self {
+                Either::Left(text.match_indices("Self"))
+            } else {
+                Either::Right(iter::empty())
+            });
```

Comment:
from what i remember `Iterator::chain` is quite slow in most cases, searching for something other than the name was something I wanted to figure out for sometime now as its also required for the `Self` type. I'll take a look at this later and see what we can do here.

Reply:
I suppose we can put the loop body in a closure for now and run that closure on both iters separately for the time being.

Reply:
updated to multiple loops. maybe we can create `&[&str]` version of the [Pattern](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html) trait to handle in one `match_indices`.